plugins/inputs/influxdb_listener/influxdb_listener.go

```diff
@@ -0,0 +1,398 @@
+				return
+			}
+			defer body.Close()
+		}
+		body = http.MaxBytesReader(res, body, h.MaxBodySize.Size)
+
+		parser := influx.NewStreamParser(body)
+		parser.SetTimeFunc(h.TimeFunc)
+
+		var err error
+		precisionStr := req.URL.Query().Get("precision")
+		if precisionStr != "" {
+			var precision time.Duration
+			precision, err = getPrecisionMultiplier(precisionStr)
+			if err != nil {
+				h.Log.Debug(err.Error())
+				badRequest(res, err.Error())
+			}
+			parser.SetTimePrecision(precision)
+		}
+
+		var m telegraf.Metric
+		var partial bool
+		for {
+			m, err = parser.Next()
+			h.bytesRecv.Incr(int64(parser.Position())) //todo incr by position difference
+
+			//continue parsing metrics even if some are malformed
+			if _, ok := err.(*influx.ParseError); ok {
+				//todo save error for X-Influxdb-Error header
+				partial = true
+				continue
+			} else if err != nil {
+				break
+			}
+
+			if h.DatabaseTag != "" && db != "" {
+				m.AddTag(h.DatabaseTag, db)
+			}
+			h.acc.AddMetric(m)
+		}
+		if err.Error() != "EOF" {
+			h.Log.Debug(err.Error())
+			// problem reading the request body
+			badRequest(res, err.Error())
+			return
+		}
+		if partial {
+			res.WriteHeader(http.StatusBadRequest)
+		} else {
+			res.WriteHeader(http.StatusNoContent)
+		}
+	}
+}
+
+func tooLarge(res http.ResponseWriter) {
+	res.Header().Set("Content-Type", "application/json")
+	res.Header().Set("X-Influxdb-Version", "1.0")
+	res.Header().Set("X-Influxdb-Error", "http: request body too large")
+	res.WriteHeader(http.StatusRequestEntityTooLarge)
+	res.Write([]byte(`{"error":"http: request body too large"}`))
+}
+
+func badRequest(res http.ResponseWriter, errString string) {
+	res.Header().Set("Content-Type", "application/json")
+	res.Header().Set("X-Influxdb-Version", "1.0")
+	if errString == "" {
+		errString = "http: bad request"
+	}
+	res.Header().Set("X-Influxdb-Error", errString)
+	res.WriteHeader(http.StatusBadRequest)
+	res.Write([]byte(fmt.Sprintf(`{"error":%q}`, errString)))
+}
+
+func partialWrite(res http.ResponseWriter, errString string) {
+	res.Header().Set("Content-Type", "application/json")
+	res.Header().Set("X-Influxdb-Version", "1.0")
+	res.Header().Set("X-Influxdb-Error", errString)
+	res.WriteHeader(http.StatusBadRequest)
+	res.Write([]byte(fmt.Sprintf(`{"error":%q}`, errString)))
+}
+
+func (h *InfluxDBListener) handleAuth(f http.HandlerFunc) http.HandlerFunc {
+	return func(res http.ResponseWriter, req *http.Request) {
+		res.Header().Set("WWW-Authenticate", "Basic realm=\"Restricted\"")
+
+		if h.BasicUsername != "" && h.BasicPassword != "" {
+			reqUsername, reqPassword, ok := req.BasicAuth()
+			if !ok ||
+				subtle.ConstantTimeCompare([]byte(reqUsername), []byte(h.BasicUsername)) != 1 ||
+				subtle.ConstantTimeCompare([]byte(reqPassword), []byte(h.BasicPassword)) != 1 {
+
+				h.authFailures.Incr(1)
+				http.Error(res, "Unauthorized", http.StatusUnauthorized)
+				return
+			}
+		}
+
+		f(res, req)
+	}
+}
```

Comment:
Use internal.AuthHandler

Reply:
Done.  I changed internal.AuthHandler to call http.Error so all users wouldn't have to do it separately.  I also added a realm parameter.